[pytorch] use cublas lt interface for bias fusion

jspark1105 opened the pull request:
Summary:
To quantify how much cublas lt interface can help param bench linear perf

On V100 GPU

for b in 512 1024; do for i in {1..3}; param_bench/train/compute/pt/pytorch_linear.py --device gpu --dtype=float16 --hidden-size 1024 --batch-size ${b}; done; done

Before this commit
batch size 512: median 21.4 TF/s (20.7, 20.6, 21.8, 21.6, 21.4)
batch size 1024: median 40.1 TF/s (39.4, 39.3, 40.2, 40.4, 40.1)

After this commit
batch size 512: median 23.5 TF/s (23.2, 23.5, 23.8, 23.9, 23.6 ) 9.8% speedup
batch size 1024: median 41.6 TF/s (42.7, 41.6, 40.4, 41.3, 41.9 ) 3.7% speedup

Differential Revision: D32053748

jspark1105 commented:
CC @xwang233

xwang233 commented:
cc @ptrblck

ngimel commented:
Hey, @xwang233, @ptrblck do you think this can be merged? It improves performance on workloads with linear + bias (which is pretty much all of them)

xwang233 commented:
Thank you @ngimel . We'll do some tests internally and get back to you soon.

xwang233 commented:
We've done some tests on performance and unit tests, and the results look good. Thanks.

NiceiceCC commented:
There is an edge case that this pr runs slower. Depends on proportion of latency from elementwise op, there could be other edge cases. 
Test setting:
ch_in: 4096*8 ch_out: 2048 bias: on dtype: float16 GPU: A100 CUDA v11.4 Driver: 470.57.0.2

With this pr, linear op runs at 152 TFLOPS. The underline kernel for mm is ampere_fp16_s16816gemm_fp16_128x128_ldg8_relu_f2f_stages_64x3_tn.

With original code, linear op runs at 192 TFLOPS. The underline kernel for mm is ampere_fp16_s16816gemm_fp16_256x128_ldg8_f2f_stages_64x3_tn

jspark1105 commented:
@xwang233 we've seen an edge case as @NiceiceCC reported. Do you expect such regression is rare only for very few edge cases?

xwang233 commented:
@jspark1105 It should be a rare corner case. You can check with the latest cuda version. We'll also check this with cublas team. Thanks!

jspark1105 committed:
[pytorch] use cublas lt interface for bias fusion (#71200)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/71200

To quantify how much cublas lt interface can help param bench (https://github.com/facebookresearch/param/) linear perf

On V100 GPU

for b in 512 1024; do for i in {1..5}; param_bench/train/compute/pt/pytorch_linear.py --device gpu --dtype=float16 --hidden-size 1024 --batch-size ${b}; done; done

Before this commit
batch size 512: median 21.4 TF/s (20.7, 20.6, 21.8, 21.6, 21.4)
batch size 1024: median 40.1 TF/s (39.4, 39.3, 40.2, 40.4, 40.1)

After this commit
batch size 512: median 23.5 TF/s (23.2, 23.5, 23.8, 23.9, 23.6 ) 9.8% speedup
batch size 1024: median 41.6 TF/s (42.7, 41.6, 40.4, 41.3, 41.9 ) 3.7% speedup

Reviewed By: ngimel

Differential Revision: D32053748

fbshipit-source-id: d95e20eb906fc60168613d2db7dd0586a4b09478
